Added more helpful message for AmbiguousParameterSet exception

kvprasoon opened the pull request:
Added more helpful message for AmbiguousParameterSet exception which always throws exception "Parameter set cannot be resolved using the specified named parameters" which is not properly conveying the end user for further steps required.

<!--

If you are a PowerShell Team member, please make sure you choose the Reviewer(s) and Assignee for your PR.
If you are not from the PowerShell Team, you can leave the fields blank and the Maintainers will choose them for you. If you are familiar with the team, feel free to mention some Reviewers yourself.

For more information about the roles of Reviewer and Assignee, refer to [CONTRIBUTING.md](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md).

-->

kvprasoon committed:
Added -Depth parameter to Split-Path, this will allow to mention the depth of the split operation.

kvprasoon committed:
Added default value 10  to Depth parameter

kvprasoon committed:
Revert "Added default value 10  to Depth parameter"

This reverts commit 128a4543df8e2939777180c77671695acc2f6233.

kvprasoon committed:
Revert "Added -Depth parameter to Split-Path, this will allow to mention the depth of the split operation."

This reverts commit 8215869e463d7c94ce5c912cf4e9707c2a06210f.

kvprasoon committed:
helpful message for Parameter set exception

Added more helpful message for AmbiguousParameterSet exception which always throws exception "Parameter set cannot be resolved using the specified named parameters" which is not properly conveying the end user for further steps required.

markekraus commented on src/System.Management.Automation/resources/ParameterBinderStrings.resx:
```diff
@@ -142,7 +142,7 @@
   </data>
   <data name="AmbiguousParameterSet" xml:space="preserve">
-    <value>Parameter set cannot be resolved using the specified named parameters.</value>
+    <value>Parameter set cannot be resolved using the specified named parameters.One or few of the parameters mentioned cannot be used together.Do a Get-command {0} to see the available parameter sets.</value>
```
* Add spaces between periods and the first word of a sentence
* `One or few` should be `One or more` but, this could also occur when an insufficient number of parameters have been supplied, not just when 2 parameters cannot be used together. 
* `parameters mentioned` should be `parameters issued`.
* `Do a Get-command {0}` do all instances of this error pass the current command name to this string for formatting? if not, this could be very confusing.
* `Get-Help` would be preferable to `Get-Command` as `Get-Command` does not show this information be default.
*  `parameter sets` is some what esoteric to function./cmdlet authors. To an admin user it might be better to say `For syntax help`.

kvprasoon replied:
Hi , Thanks for the review, I'll do the required modifications

markekraus reviewed (changes requested):
Thanks for contributing! Please make modifications to your pull request.

kvprasoon committed:
Added more helpful message for AmbiguousParameterSet exception



Previoues review comments
• Add spaces between periods and the first word of a sentence - Done
• One or few  should be  One or more  but, this could also occur when an insufficient number of parameters have been supplied, not just when 2 parameters cannot be used together. - Done
• parameters mentioned  should be  parameters issued . - Done
• Do a Get-command {0}  do all instances of this error pass the current command name to this string for formatting? if not, this could be very confusing. - Done, changed to a generic message without deending on the first element
• Get-Help  would be preferable to  Get-Command  as  Get-Command  does not show this information be default. - Done, changed to Get-Help
• parameter sets  is some what esoteric to function./cmdlet authors. To an admin user it might be better to say  For syntax help . - Done

kvprasoon committed:
Corrrecting the Intent issues @ line 113 and 123

kvprasoon commented:
I have updated it as per the comments

kvprasoon committed:
added 'cmdlet' to the error message

Added cmdlets along with script/function in the error message .

kvprasoon reviewed (commented):
Thanks for the review,I have updated it as per the comments

• Add spaces between periods and the first word of a sentence - Done
• One or few  should be  One or more  but, this could also occur when an insufficient number of parameters have been supplied, not just when 2 parameters cannot be used together. - Done
• parameters mentioned  should be  parameters issued . - Done
• Do a Get-command {0}  do all instances of this error pass the current command name to this string for formatting? if not, this could be very confusing. - Done, changed to a generic message without deending on the first element
• Get-Help  would be preferable to  Get-Command  as  Get-Command  does not show this information be default. - Done, changed to Get-Help
• parameter sets  is some what esoteric to function./cmdlet authors. To an admin user it might be better to say  For syntax help . - Done

kvprasoon reviewed (commented):
Thanks for the review,I have updated it as per the comments

• Add spaces between periods and the first word of a sentence - Done
• One or few  should be  One or more  but, this could also occur when an insufficient number of parameters have been supplied, not just when 2 parameters cannot be used together. - Done
• parameters mentioned  should be  parameters issued . - Done
• Do a Get-command {0}  do all instances of this error pass the current command name to this string for formatting? if not, this could be very confusing. - Done, changed to a generic message without deending on the first element
• Get-Help  would be preferable to  Get-Command  as  Get-Command  does not show this information be default. - Done, changed to Get-Help
• parameter sets  is some what esoteric to function./cmdlet authors. To an admin user it might be better to say  For syntax help . - Done

markekraus commented on src/System.Management.Automation/resources/ParameterBinderStrings.resx:
```diff
@@ -142,7 +142,7 @@
   </data>
   <data name="AmbiguousParameterSet" xml:space="preserve">
-    <value>Parameter set cannot be resolved using the specified named parameters.</value>
+    <value>Parameter set cannot be resolved using the specified named parameters. One or more of the parameters issued cannot be used together. Do a Get-Help on your script/function/cmdlet for the syntax help.</value>
```
* `One or more of the parameters issued cannot be used together.` This still doesn't address the case where too few parameters are supplied. For example: `foo` has 2 parameter sets `-A -B -C` and `-A -B -D` but the user only supplied `-A -B`. It's not that `-A` and `-B` can't be used together, but that `-C` or `-D` are needed to determine the parameter set.
* `Do a Get-Help on your script/function/cmdlet for the syntax help.`  I think this is would be somewhat confusing. I'm not sure how to fix it. The wording could at least be improved `Use Get-Help with your script, function, or cmdlet for syntax help.`

kvprasoon committed:
Added more infor as pe rthe review comment

One or more of the parameters issued cannot be used together.  This still doesn't address the case where too few parameters are supplied. For example:  foo  has 2 parameter sets  -A -B -C  and  -A -B -D  but the user only supplied  -A -B . It's not that  -A  and  -B  can't be used together, but that  -C  or  -D  are needed to determine the parameter set. - Added the information about not using the decision making parameter
• Do a Get-Help on your script/function/cmdlet for the syntax help.  I think this is would be somewhat confusing. I'm not sure how to fix it. The wording could at least be improved  Use Get-Help with your script, function, or cmdlet for syntax help.  - changed to suggested message

kvprasoon reviewed (commented):
One or more of the parameters issued cannot be used together.  This still doesn't address the case where too few parameters are supplied. For example:  foo  has 2 parameter sets  -A -B -C  and  -A -B -D  but the user only supplied  -A -B . It's not that  -A  and  -B  can't be used together, but that  -C  or  -D  are needed to determine the parameter set. - Added the information about not using the decision making parameter
• Do a Get-Help on your script/function/cmdlet for the syntax help.  I think this is would be somewhat confusing. I'm not sure how to fix it. The wording could at least be improved  Use Get-Help with your script, function, or cmdlet for syntax help.  - changed to suggested message

kvprasoon reviewed (commented):
please review

kvprasoon reviewed (commented):
can anyone help me by reviewing this

iSazonov commented:
@kvprasoon I guess PowerShell team is fully obsessed with the release of the new version.
Also please rebase to get latest updates - I think this help to pass CIs.

/cc @SteveL-MSFT  Should we use direct tips in error messages or use HelpSystem capabilites?

markekraus commented on src/System.Management.Automation/resources/ParameterBinderStrings.resx:
```diff
@@ -142,7 +142,7 @@
   </data>
   <data name="AmbiguousParameterSet" xml:space="preserve">
-    <value>Parameter set cannot be resolved using the specified named parameters.</value>
+    <value>Parameter set cannot be resolved using the specified named parameters. One or more parameters issued cannot be used together or a parameter which will decide the parameter set is not provided. Use Get-Help with your script, function, or cmdlet for syntax help.</value>
```
> or a parameter which will decide the parameter set is not provided.

Perhaps this should be `or an insufficient number of parameters were provided.`  There could certainly be more than one missing.

SteveL-MSFT commented:
@iSazonov since you added `Type 'help' for help` as part of the startup banner, it seems unnecessary to point users to `get-help` in any error message.  PowerShell does have a suggestion system that we should try to take advantage of in the future to provide more concise help on how to resolve issues (like if they type a parameter name wrong, we can do some matching and suggest what they meant which is different from this PR).

For this particular issue, I think the ideal error message would tell the user which parameters are mutually exclusive and can't be used together or if another parameter is required to resolve the parameter set.  I added a note to https://github.com/PowerShell/PowerShell/issues/5175 which covers a new proposal on defining mutually exclusive parameters to have better error messages.

For this PR, I think it would be sufficient to update the error message to what @markekraus is proposing without the `get-help` mention on the end which is to generic.

iSazonov commented:
@SteveL-MSFT Thanks for clarify and confirm my thoughts.

markekraus commented:
#5435 has been opened instead.
